{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b8e3ae3-c6d7-443c-98c2-1d8b9871e84a",
   "metadata": {},
   "source": [
    "# Chapter 21 – DNA Sequence Analysis of MinION Nanopore Reads\n",
    "Our project is dedicated to analyzing sequence reads from a plasmid following its digestion by either a restriction enzyme or a Cas9/gRNA complex. The aim is the identification of cutting sites within the plasmid. The reads were generated by DNA sequencing with a MinION, a pocket-sized sequencing device based on Nanopore technology. For the project, the FASTA5 read files are provided. The binary FASTA5 format standardizes nucleotide sequence data storage, ensuring compatibility across various bioinformatics tools. For the analysis, a suite of specialized tools is employed. The Guppy tool facilitates base calling and demultiplexing of nanopore sequence reads based on barcode identifiers. qfilter, a little tool developed in my lab, is utilized to filter sequencing data, retaining only high-quality reads while discarding noisy or low-quality data. The Minimap tool aligns long nucleotide sequences to a reference genome, facilitating the detection of similarities and diﬀerences relevant to cutting sites. This in turn is done with the Integrated Genome Viewer (IGV). Overall, the combined use of these tools enables the processing and analysis of MinION-derived sequencing data, ultimately leading to the identification of sequence features. Finally, the simulation of nanopore sequence reads is exemplified. \n",
    "\n",
    "> If you read these instructions on **GitHub**, you will notice that some lines of code are incomplete at the end. Nevertheless, you can copy the cells completely into the clipboard by double-clicking on them.\n",
    "\n",
    "Necessary Installations:\n",
    "\n",
    "- `sudo apt install -y hdfview hdf5-tools ont-fast5-api dotter porechop minimap2 samtools igv`\n",
    "- `wget https://cdn.oxfordnanoportal.com/software/analysis/ont_guppy_cpu_6.3.7-1~focal_amd64.deb`\n",
    "- `sudo apt install -f ./ont_guppy_cpu_6.3.7-1~focal_amd64.deb`\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f671cb83-444f-4863-9840-b6a4c563e8ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bac98e1d-5ca2-4b05-b94d-e8f8448acc4d",
   "metadata": {},
   "source": [
    "### Download Data Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed444e88-8613-4193-9e2e-9c0ed9d929f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "wget https://www.dropbox.com/s/si08bjwdvcmtuls/ALT327_c9ffe0f3_0.fast5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10811370-6484-4fea-b96f-9e0c493e993a",
   "metadata": {},
   "outputs": [],
   "source": [
    "wget https://www.dropbox.com/s/88hqohg3271un21/pBR322.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fc4aab9-647d-46f8-97d5-62c8185afce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2571d3a-8130-4681-92b1-9a2f87775cbc",
   "metadata": {},
   "source": [
    "### Move Files to new Folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aee75c3-f12e-47d2-9aec-bd6694fca6d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir OriginalFast5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c25a50e-51a1-462a-9e05-af0806a333fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "cp ALT327_c9ffe0f3_0.fast5 OriginalFast5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8bf4391-3ef3-4f87-8857-1a11901581aa",
   "metadata": {},
   "source": [
    "### Viewing the FAST5 Files\n",
    "FAST5 files are Hierarchical Data Format 5 (HDF5) files with a specific structure defined by Oxford Nanopore Technologies (ONT) for the storage of raw sequencing data and are generated by, for example, MinION. There are two types of FAST5: Single-FAST5 and Multi-FAST5. A multi-FAST5 file contains data for several reads in a single file, while a single-FAST5 file contains data from a single read. Single FAST5 files are rarely used.\n",
    "A free program for viewing the files is HDFView, which you can download for all operating systems at [hdfgroup.org](https://hdfgroup.org). We have already installed it above with APT.\n",
    "\n",
    "\n",
    "You can start the graphical version with `hdfview` in the terminal.\n",
    "\n",
    "However, we use command line tools here.\n",
    "\n",
    "The program `h5ls` is available for this purpose, which lists the contents of a FAST5 file. As the MinION file is a multi-FAST5 file, we limit the output to three lines with the command `head -3`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1f8a5ba-4d20-47ac-9e71-711572b8809a",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ls OriginalFast5/ALT327_c9ffe0f3_0.fast5 | head -3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "142c7657-0abc-404b-aa59-ed0417a68c6a",
   "metadata": {},
   "source": [
    "To check what is stored for a single read, you can specify this read with `h5ls` like a directory, i.e. separated with a slash:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76ec1e65-3430-4c1e-b5a7-79687802bf12",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ls OriginalFast5/ALT327_c9ffe0f3_0.fast5/read_0048d23d-4b49-476b-b18c-d102e7a3e393"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3a83fd8-eaa0-4132-aad3-d7ddf0b86034",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ls OriginalFast5/ALT327_c9ffe0f3_0.fast5/read_0048d23d-4b49-476b-b18c-d102e7a3e393/Raw/Signal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25567396-cfb2-4988-b23b-265f0df98400",
   "metadata": {},
   "source": [
    "With the `--data` option we can display individual data records. A “subdirectory” contains the raw data from the MinION, for example the voltage changes across a pore (squiggle plot):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2332611-96c1-4b4c-8af9-1ad0929622a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ls -d OriginalFast5/ALT327_c9ffe0f3_0.fast5/read_0048d23d-4b49-476b-b18c-d102e7a3e393/Raw/Signal | head -3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "498cc15f-f8d9-49fb-9ce6-762f7b04c377",
   "metadata": {},
   "source": [
    "**Oops**: *\"Unable to print data.\"* Something went wrong!?! Oxford Nanopore uses two different methods to compress the raw data in the FAST5 files: vbz or gzip. Recently, ONT's proprietary VBZ compression has been used, which older analysis programs sometimes have problems with. The program suite `ont-fast5-api` can be used to convert the compression and thus solve the problem of unreadable data (see above). We have already completed the installation above.\n",
    "\n",
    "The data compression is then converted, where `--input_path OriginalFast5` names the directory with the FAST5 files and we use decompressed to name the output directory. Then the data extraction works:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcd5bb7d-801d-4375-b183-52ce37347f8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "compress_fast5 --input_path OriginalFast5 --save_path OriginalFast5Gzip --compression gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d3a7903-3bdc-4617-9ef0-338cf7f11c12",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ls -d OriginalFast5Gzip/ALT327_c9ffe0f3_0.fast5/read_0048d23d-4b49-476b-b18c-d102e7a3e393/Raw/Signal | head -3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0b13ad1-a4d6-4f3e-a1ee-7dd4a9412cc9",
   "metadata": {},
   "source": [
    "**Tip**: To speed up data analysis, the thousand reads in the FAST5 file can be separated into individual FAST5 files. This is also possible with the program suite `ont-fast5-api`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50061006-0c61-4834-9dc1-ec1acc757af9",
   "metadata": {},
   "outputs": [],
   "source": [
    "multi_to_single_fast5 --input OriginalFast5Gzip --save OriginalFast5GzipSingle"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b99ccf11-5bfd-4159-9c6f-ec0734a9517b",
   "metadata": {},
   "source": [
    "*OriginalFast5Gzip* denotes the directory with the FAST5 files and *OriginalFast5GzipSingle* denotes the output directory. FAST5 files can then be created with, for example, one hundred reads each:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "115debfb-6882-430b-97ad-8f5cb6d896a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "single_to_multi_fast5 --input OriginalFast5GzipSingle/0 --save OriginalFast5GzipMulti100 --batch_size 100"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8b51979-afc1-4fd4-878b-52e02fa11b6f",
   "metadata": {},
   "source": [
    "The `--save` option denotes the output directory and the `--batch_size` option denotes the number of reads per file, each of which is named *batch_0.fast5*, *batch_1.fast5*, *batch_2.fast5*, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd4fa1b3-1d84-4f14-9e28-43290b73334e",
   "metadata": {},
   "source": [
    "### Basecalling with Guppy\n",
    "Guppy is a program that includes Oxford Nanopore Technologies' basecalling algorithms and several post-processing bioinformatics applications. This also includes demultiplexing, adapter trimming and alignments with target sequences, such as the plasmid pBR322. It is available for Windows, MacOS and Linux operating systems from Oxford Nanopore. The installation is described above. In addition to the path to a directory with our FAST5 file(s) (option `--input_path`), the basecaller needs the name of an output directory (`--save_path`) and the flow cell (`--flowcell`) and sequencing kit used - (`--kit`) combination, in our case the Flongle (FLO-FLG001) and the Rapid Barcoding Kit (SQK-RBK004). The output directory will be created automatically if it does not already exist.\n",
    "\n",
    "On my Linux virtual machine in VitualBox on a MacBook Pro, basecalling a thousand sequences takes around an hour with one CPU (3,478,643 ms is around 58 min).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be20f8cc-8f07-4ec5-aa3b-189f47a131e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "guppy_basecaller --input_path OriginalFast5 --flowcell FLO-FLG001 --kit SQK-RBK004 --save_path BasecalledAll"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c37ce0ed-9743-43ca-b5b7-2d55f751f361",
   "metadata": {},
   "source": [
    "### A first analysis with yolk\n",
    "In the *basecalled/pass* directory there is a file in FASTQ format that contains the DNA sequences of the reads. Actually, we can only really talk about reads now. We will now convert this file into a file in FASTA format. To do this, we use the AWK programming language, which can be used to easily format text. Here we instruct awk to output every fourth line from line 1 or line 2. With `head -10` we limit the output to ten lines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a5b5bc6-b947-4ffb-b78d-323f76e9966e",
   "metadata": {},
   "outputs": [],
   "source": [
    "awk 'NR%4==1 {print \">\"$1} NR%4==2 {print $0}' BasecalledAll/pass/fastq_runid_c9ffe0f31e537612d2cbf4c0f7e050d9b0d928a0_0_0.fastq | head -10 > rohreads10.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5bc9919-cf83-4aee-8d2b-ccf1889afe2e",
   "metadata": {},
   "source": [
    "Dotter is a graphical dotplot program for the detailed comparison of two DNA sequences. This involves comparing every nucleotide in one sequence with every nucleotide in the other sequence. The first sequence is along the X-axis and the second sequence (or more) is along the Y-axis. In areas where the two sequences are similar, a diagonal runs through the matrix. If you were to compare a sequence to itself, you would get an angle bisector. The program call is followed by the FASTA files of the reference and the sequences to be compared.\n",
    "\n",
    "You start Dotter with `dotter pBR322.fasta rohreads10.fasta` in the command line. The terminal is now no longer ready for input - the input prompt ($) only appears again after Dotter is closed. We receive a dot plot in a graphical window, which gives us a first impression of our result.\n",
    "\n",
    "![dotter](dotter.png))\n",
    "\n",
    "The sequence of the plasmid is plotted on the x-axis and the sequence of the reads is plotted on the y-axes, separated by green lines. If nucleotides on the X and Y axes match, then a point is drawn; hence the name dotplot."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eb337e1-dcaa-4e6e-927d-fa6ef4613085",
   "metadata": {},
   "source": [
    "### Demultiplexing with PoreChop\n",
    "When different experiments are carried out, each separated by barcodes and sequenced on a flow cell, all reads end up in a FAST5 or FASTQ file. These can then be separated from each other using the `porechop` program."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e4c2470-23ca-4f49-bb1e-f73c1356d1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "porechop --input BasecalledAll/pass/fastq_runid_c9ffe0f31e537612d2cbf4c0f7e050d9b0d928a0_0_0.fastq --barcode_dir BarCodesAll"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ed1b955-496a-439a-88e0-1f8f16ab2761",
   "metadata": {},
   "source": [
    "The `--barcode_dir` option specifies the directory in which the found barcodes are saved. The entire issue is quite long. The last lines are crucial and show how many reads could be assigned to barcodes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27f81315-9519-43b8-985d-e0a6bc8797d6",
   "metadata": {},
   "source": [
    "### ALTERNATIVE with Guppy\n",
    "Guppy also has its own demultiplexer. Let us use it as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da90ba6f-1ae6-4cdd-8079-753807476ff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "guppy_barcoder --input_path BasecalledAll/pass --save_path BarcodesAll --barcode_kits SQK-RBK004"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b440422-744a-45b7-a187-892d25a1ce7c",
   "metadata": {},
   "source": [
    "Let us count the number of demultiplexed reads:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea25df10-ba89-4884-a849-1b81cb945e41",
   "metadata": {},
   "outputs": [],
   "source": [
    "grep -c \"^@\" BarcodesAll/barcode0?/*.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a99f4272-d9f5-4257-981c-3b2336e003b7",
   "metadata": {},
   "source": [
    "### Quality control and filtering of reads\n",
    "The quality of the reads produced by the MinION will vary. If the quality is too poor, a match to the plasmid sequence cannot be detected later. The quality is indicated with the so-called PHRED quality score (or simply quality score). It has a logarithmic relationship to the base call error probabilities in base calling. A quality score of 10 means that 90% of the bases are correct; with a value of 20 it is 99%.\n",
    "In addition to quality, we also filter based on read length. When the sequencing library is created, DNA molecules can attach to one another: so-called concatemers are created.\n",
    " \n",
    "First we download the qFilter program and make it executable:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e57625c-5ad8-4a11-acbd-9ab570ae81c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "wget \"https://github.com/awkologist/qfilter/raw/main/qfilter\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "805cf150-831f-40ad-b4e7-67987f67ae9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "chmod u+x qfilter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38c37c25-4da1-4a36-a167-c2743ce4f897",
   "metadata": {},
   "source": [
    "We then filter the reads. Here we use a for loop with which all three files *BC01.fastq*, *BC02.fastq* and *BC03.fastq* are processed one after the other. The for loop has the format: `for` variable `in` list`; do` command(s)`; done` We use *i* as a variable, which takes on the values one to three one after the other. The variable value is called up again with *$i*.\n",
    "\n",
    "With `-m` we specify the minimum average PHRED score of a read, with `-x` the minimum read length and with `-y` the maximum read length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5bf2780-2adf-4768-a640-000286b96736",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in 1 2 3; do ./qfilter -m 10 -x 1000 -y 4361 BarCodesAll/BC0$i.fastq > bc0$i.all.fastq ;done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3e01f8a-e00d-4fad-9d61-b863c1d10541",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc -l bc0[123].all.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e87e75cb-7245-4494-8414-887324ad5ae9",
   "metadata": {},
   "source": [
    "### ALTERNATIVE with Guppy Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d079f7e8-0f19-4906-ae73-ea49a280e339",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in 1 2 3; do ./qfilter -m 10 -x 1000 -y 4361 BarcodesAll/barcode0$i/*.fastq > bc0$i.all.fastq ;done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e6df87b-4c6b-406f-8c57-f3a3150c6b26",
   "metadata": {},
   "outputs": [],
   "source": [
    "grep -c \"^@\" bc0[123].all.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "605f81ea-6b15-4804-b565-3d30c67b65da",
   "metadata": {},
   "source": [
    "### Mapping the reads to the plasmid\n",
    "We use the Minimap program to map the reads to the plasmid pBR322. “Mapping” refers to the process that looks at which positions in the plasmid sequence the reads generated by the MinION fit. A sequence alignment is therefore carried out with each read. Here again we use a for loop, in this case with file names as a list; The question mark stands as a joker for any character."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2d32c61-165e-4783-afd7-5dcc32ad4af5",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in bc0?.all.fastq; do minimap2 -a pBR322.fasta $i > $i.mapped.sam; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df596fed-5eb2-403c-8e37-40048e105bc5",
   "metadata": {},
   "source": [
    "The `-a` option tells Minimap2 to output the result in the so-called SAM format. Normally the output would be to the screen. I cropped the output again. With the redirection `>` we redirect the data to a file that we name *mapped.sam*."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efedf045-8c7c-450f-853d-b43278f5893a",
   "metadata": {},
   "source": [
    "### ALTERNATIVE with Guppy Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b7bb5fe-b815-4f91-b438-07dddb43145b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in bc0?.all.guppy.fastq; do minimap2 -a pBR322.fasta $i > $i.mapped.guppy.sam; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce47c3f8-c9ba-4260-9da9-6462a1d541fe",
   "metadata": {},
   "source": [
    "### Preparation for Visualization with IGV\n",
    "The Integrated Genome Viewer (IGV) is a program for viewing the mapped reads. To do this, the reference is first loaded (genome) and then the reads (track). In our case, the genome is the plasmid pBR322 in the file *pBR322.fasta* and the mapped reads are in the file *mapped.sam* previously created using Mminimap2. This still needs to be sorted, indexed and converted into binary format for IGV. We do this with the program `samtools`. Again we use the For loop to sort all three files, convert them to binary BAM format and then index them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0b4960f-80a9-4d00-85ea-41f80eaab747",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in bc0?.all.fastq.mapped.sam; do samtools sort $i -o $i.sorted.bam; samtools index $i.sorted.bam; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a593e6e-efc8-4b5b-a0c1-10d9d796bb20",
   "metadata": {},
   "source": [
    "The result is the sorted files *bc0[123].all.fastq.mapped.sorted.bam* with their index in the file *bc0[123].all.fastq.mapped.sorted.bam.bai*, the IGV in background will load. With \"[123]\" I want to indicate that there are three files each.\n",
    "The plasmid (genome) automatically indexes IGV when loading :-)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c56ad47c-3a3e-4d4b-a546-7c57d81223d1",
   "metadata": {},
   "source": [
    "### ALTERNATIVE with Guppy Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57e17643-3ed7-4253-bf9c-8bdc78d39ea6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in bc0?.all.guppy.fastq.mapped.guppy.sam; do samtools sort $i -o $i.sorted.guppy.bam; samtools index $i.sorted.guppy.bam; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1da8fcf1-4a62-43f7-8e5d-1db52227caa6",
   "metadata": {},
   "source": [
    "### Visualization with IGV\n",
    "The Integrated Genome Viewer(`igv`) is a versatile JAVA program for displaying genomic data \n",
    "After starting the program, the terminal is occupied like with Dotter and can only be fed with commands again when the IGV window has been closed.\n",
    "\n",
    "In the IGV window we first specify the reference genome in the menu item **Genomes > Load Genomes from File ...**. In our humble case it is the plasmid pBR322, the sequence of which is stored in the FASTA file *pBR322.fasta*. We then load the reads of the three experiments in BAM format (*bc0[123].all.fastq.mapped.ont.sam.sorted.bam*) in the menu **File > Load from File ...** With each When you load a data record, the window fills up. For each BAM file (or barcode or experiment) we see the reads and, above them, the hit density (coverage) that these reads generate. Don't be surprised if things look a little different for you.\n",
    "\n",
    "![igv](igv.png)\n",
    "\n",
    "It is noticeable that in the upper data set we see a gap at 670 base pairs. In the middle data set there is a gap at around 3,760 base pairs. In the lower data set, however, the entire plasmid sequence is covered by reads. \n",
    "What does that tell us? In the\n",
    "\n",
    "- upper data set was the plasmid was cut once,\n",
    "- twice in the middle data set (Cas9 and PstI),\n",
    "- and in the lower data set not cut at all (undigested control).\n",
    "\n",
    "Remember that reads cannot lie across interfaces. In the upper data set we find an exception: one read spans the interface at 670 bp, which may indicate a partially incomplete digestion\n",
    "In the analysis presented, we only used the first thousand reads, which were saved by the MinION on the connected computer after around ten minutes of sequencing. That's a fraction of all data."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
